### ISSUE TITLE
Add query parameter to all lists used by the Wazuh app

### ISSUE THREAD

**Issue by jesusgn90** [wazuh/wazuh-api#248]
Hi team, we need `q` for the next routes:

- [x] /agents/groups
- [x] /cluster/:node/logs
- [x] /manager/logs
- [x] /cluster/nodes
- [x] /rules
- [x] /decoders
- [x] /syscheck/:agent

Maybe I miss some other list, but the above routes must fit this requirement due to https://github.com/wazuh/wazuh-kibana-app/issues/1002, thanks in advance.

**Comment by druizz90** [comment]
Hi team,

For adding `q` parameter to `/agents/groups` call we have `get_all_groups_sql` method, which is unused actually and we could use it ([get_all_groups_sql](https://github.com/wazuh/wazuh/blob/8fb9bdfcfc6acfb4b9fc60b095071136be3d8bbe/framework/wazuh/agent.py#L1270-L1285)).

With `/syscheck/:agent_id` we could do a class which inherits from `WazuhDBQuery` to replace [this](https://github.com/wazuh/wazuh/blob/8fb9bdfcfc6acfb4b9fc60b095071136be3d8bbe/framework/wazuh/syscheck.py#L146-L148).

For the others, we don't use queries to the database. We could create a database for rules (for example) or do something like as SQL query when we retrieve the data (it may be too slow).

Thoughts?

**Comment by mgmacias95** [comment]
Hello,

A first approach adding support for CIS-CAT and Syscollector requests has been developed here https://github.com/wazuh/wazuh/pull/2973.

Best regards,
Marta

**Comment by crd1985** [comment]
This issue is blocked by PR https://github.com/wazuh/wazuh/pull/2973.

**Comment by druizz90** [comment]
# Issue update

We need to make a function for searching in arrays or dictionaries as @mgmacias95 said in https://github.com/wazuh/wazuh-api/issues/248#issuecomment-444453346 in order to add `q` parameter to these endpoints.

`/agents/groups`, `/cluster/:node/logs`, `/manager/logs` and `/cluster/nodes` not get their data only from a database request and it is for that by we need to create a new function. This approach could be enough for our purpose with `rules` and `decoders` too.

`/syscheck/:agent` is supporting `q` parameter on branches `dev-queries-ciscat-syscollector-361` (API) and `dev-queries-ciscat-syscollector-2973`:
```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/syscheck/000?pretty&q=size=37736"
{
   "error": 0,
   "data": {
      "items": [
         {
            "perm": "100755",
            "file": "/usr/bin/cal",
            "gname": "root",
            "date": "2019-07-19 06:05:29",
            "gid": "0",
            "uname": "root",
            "mtime": "2018-10-30 22:48:52",
            "size": 37736,
            "uid": "0",
            "sha1": "d6bf46f589d22fd263c662514dd9fa7d4db48b3a",
            "type": "file",
            "inode": 2630165,
            "sha256": "f0ad9bea071e7584a62e81a74484025bd361c87da2fa93e8869935bf75ac3476",
            "md5": "f368add3f22bd9c4b58b57ce9e0de7d4"
         }
      ],
      "totalItems": 1
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/syscheck/005?pretty&file=/usr/bin/col"
{
   "error": 0,
   "data": {
      "items": [
         {
            "gid": "0",
            "uname": "root",
            "mtime": "2018-10-30 22:48:52",
            "uid": "0",
            "gname": "root",
            "type": "file",
            "perm": "100755",
            "date": "2019-07-19 06:05:37",
            "inode": 2630186,
            "sha256": "ea526485f6bc71f023a6f4bb30ba956f26965c204aeb1c4ff55c7739abb93011",
            "md5": "d55320225ec4dd4294ff46ec2c9b1932",
            "size": 24504,
            "file": "/usr/bin/col",
            "sha1": "9276ebbeac7d4bc8d55a5780300fdb9e563dc720"
         }
      ],
      "totalItems": 1
   }
}
```

**Comment by druizz90** [comment]
# Issue update

## Summary

I am working on branches `dev-add-query-parameters-all-lists-248` (API and framework) for this development.

I added a new function to framework in `utils.py` for giving support to `q` parameter. `/agents/groups` endpoint has support for `q` parameter now. It is necessary to add the unit tests for finishing this endpoint. Mocha tests are working properly:

```javascript
GET/agents/groups
      ✓ Request (250ms)
      ✓ Retrieve all elements with limit=0 (255ms)
      ✓ Hash algorithm (283ms)
      ✓ Wrong Hash algorithm (256ms)
      ✓ Filters: query (operator =) (263ms)
      ✓ Filters: query (operator !=) (252ms)
      ✓ Filters: query (operator <) (262ms)
      ✓ Filters: query (operator >) (257ms)
      ✓ Filters: query (operator ~) (257ms)
      ✓ Filters: query (AND operations) (259ms)
      ✓ Filters: query (OR operations) (256ms)
      ✓ Filters: query (AND and OR operations) (257ms)
```

Some examples with this parameter working are the next:

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/groups?pretty&q=name~test"
{
   "error": 0,
   "data": {
      "items": [
         {
            "count": 0,
            "name": "testsagentconf",
            "mergedSum": "2acdb385658097abb9528aa5ec18c490",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         },
         {
            "count": 0,
            "name": "testsagentconf2",
            "mergedSum": "391ae29c1b0355c610f45bf133d5ea55",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         }
      ],
      "totalItems": 2
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/groups?pretty&q=count=3"  
{
   "error": 0,
   "data": {
      "items": [
         {
            "count": 3,
            "name": "default",
            "mergedSum": "a7d19a28cd5591eade763e852248197b",
            "configSum": "ab73af41699f13fdd81903b5f23d8d00"
         }
      ],
      "totalItems": 1
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/groups?pretty&q=count!=3"
{
   "error": 0,
   "data": {
      "items": [
         {
            "count": 0,
            "name": "dmz",
            "mergedSum": "dd77862c4a41ae1b3854d67143f3d3e4",
            "configSum": "ab73af41699f13fdd81903b5f23d8d00"
         },
         {
            "count": 0,
            "name": "testsagentconf",
            "mergedSum": "2acdb385658097abb9528aa5ec18c490",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         },
         {
            "count": 0,
            "name": "testsagentconf2",
            "mergedSum": "391ae29c1b0355c610f45bf133d5ea55",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         }
      ],
      "totalItems": 3
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/groups?pretty&q=mergedSum=391ae29c1b0355c610f45bf133d5ea55"
{
   "error": 0,
   "data": {
      "items": [
         {
            "count": 0,
            "name": "testsagentconf2",
            "mergedSum": "391ae29c1b0355c610f45bf133d5ea55",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         }
      ],
      "totalItems": 1
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/groups?pretty&q=mergedSum=391ae29c1b0355c610f45bf133d5ea55;name~test"
{
   "error": 0,
   "data": {
      "items": [
         {
            "count": 0,
            "name": "testsagentconf2",
            "mergedSum": "391ae29c1b0355c610f45bf133d5ea55",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         }
      ],
      "totalItems": 1
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/agents/groups?pretty&q=mergedSum=391ae29c1b0355c610f45bf133d5ea55,name~test"
{
   "error": 0,
   "data": {
      "items": [
         {
            "count": 0,
            "name": "testsagentconf",
            "mergedSum": "2acdb385658097abb9528aa5ec18c490",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         },
         {
            "count": 0,
            "name": "testsagentconf2",
            "mergedSum": "391ae29c1b0355c610f45bf133d5ea55",
            "configSum": "297b4cea942e0b7d2d9c59f9433e3e97"
         }
      ],
      "totalItems": 2
   }
}

```

`/syscheck/:agent` is supporting `q` parameter:

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/syscheck/000?pretty&q=sha1=d6bf46f589d22fd263c662514dd9fa7d4db48b3a"
{
   "error": 0,
   "data": {
      "items": [
         {
            "perm": "100755",
            "uid": "0",
            "type": "file",
            "sha1": "d6bf46f589d22fd263c662514dd9fa7d4db48b3a",
            "gname": "root",
            "size": 37736,
            "uname": "root",
            "md5": "f368add3f22bd9c4b58b57ce9e0de7d4",
            "inode": 2630165,
            "gid": "0",
            "date": "2019-07-22 11:53:25",
            "file": "/usr/bin/cal",
            "sha256": "f0ad9bea071e7584a62e81a74484025bd361c87da2fa93e8869935bf75ac3476",
            "mtime": "2018-10-30 22:48:52"
         }
      ],
      "totalItems": 1
   }
}
```

## TO DO

The followings endpoints need to be adapted for filtering by `q` parameter:

- `/cluster/:node/logs`
- `/manager/logs`
- `/cluster/nodes`
- `rules`
- `decoders`

**Comment by druizz90** [comment]
# Issue update

## Summary

`q` parameter was added to `/cluster/:node/logs` endpoint:

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/cluster/master/logs?pretty&q=tag=ossec-execd" 
{
   "error": 0,
   "data": {
      "items": [
         {
            "timestamp": "2019-07-23 08:05:15",
            "tag": "ossec-execd",
            "level": "info",
            "description": " Started (pid: 128)."
         },
         {
            "timestamp": "2019-07-23 07:40:18",
            "tag": "ossec-execd",
            "level": "info",
            "description": " Started (pid: 123)."
         },
         {
            "timestamp": "2019-07-23 07:37:50",
            "tag": "ossec-execd",
            "level": "info",
            "description": " Started (pid: 12068)."
         }
      ],
      "totalItems": 3
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/cluster/master/logs?pretty&q=tag=ossec-rootcheck&limit=2"
{
   "error": 0,
   "data": {
      "items": [
         {
            "timestamp": "2019-07-23 08:06:01",
            "tag": "ossec-rootcheck",
            "level": "info",
            "description": " Ending rootcheck scan."
         },
         {
            "timestamp": "2019-07-23 08:05:34",
            "tag": "ossec-rootcheck",
            "level": "info",
            "description": " Starting rootcheck scan."
         }
      ],
      "totalItems": 6
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/cluster/master/logs?pretty&q=tag=ossec-analysisd&limit=4"
{
   "error": 0,
   "data": {
      "items": [
         {
            "timestamp": "2019-07-23 08:05:19",
            "tag": "ossec-analysisd",
            "level": "info",
            "description": " Connected to '/queue/alerts/ar' (active-response queue)"
         },
         {
            "timestamp": "2019-07-23 08:05:16",
            "tag": "ossec-analysisd",
            "level": "info",
            "description": " Total rules enabled: '3395'"
         },
         {
            "timestamp": "2019-07-23 08:05:16",
            "tag": "ossec-analysisd",
            "level": "info",
            "description": " Ignoring file: '/etc/mtab'"
         },
         {
            "timestamp": "2019-07-23 08:05:16",
            "tag": "ossec-analysisd",
            "level": "info",
            "description": " Ignoring file: '/etc/hosts.deny'"
         }
      ],
      "totalItems": 71
   }
}
```

**Comment by druizz90** [comment]
# Issue update

## Summary

`q` parameter was added to all requested endpoints. This PR should be merged when unit tests run properly on `3.11` version.

### `decoders`

`q` parameter is working for decoders:

```bash
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/decoders?pretty&q=file~0450;position<3"
{
   "error": 0,
   "data": {
      "items": [
         {
            "file": "0450-openvas_decoders.xml",
            "path": "ruleset/decoders",
            "name": "gsad",
            "position": 0,
            "status": "enabled",
            "details": {
               "program_name": "^gsad"
            }
         },
         {
            "file": "0450-openvas_decoders.xml",
            "path": "ruleset/decoders",
            "name": "gsad",
            "position": 1,
            "status": "enabled",
            "details": {
               "prematch": "^gsad\\s*main:\\s*\\w+:\\w+|^gsad\\s*omp:\\s*\\w+:\\w+|^gsad\\s+xslt:\\s+\\w+:\\w+|^gsad\\s+base:\\s+\\w+:\\w+|^gsad\\s+vali:\\s+\\w+:\\w+"
            }
         },
         {
            "file": "0450-openvas_decoders.xml",
            "path": "ruleset/decoders",
            "name": "gsad-auth-failed",
            "position": 2,
            "status": "enabled",
            "details": {
               "parent": "gsad",
               "prematch": "Authentication failure for '",
               "regex": "^(\\.+)' from (\\S+)",
               "order": "user, srcip"
            }
         }
      ],
      "totalItems": 3
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/decoders?pretty&q=file~0320;position>10"
{
   "error": 0,
   "data": {
      "items": [
         {
            "file": "0320-sudo_decoders.xml",
            "path": "ruleset/decoders",
            "name": "sudo2-fields",
            "position": 11,
            "status": "enabled",
            "details": {
               "parent": "sudo",
               "regex": "USER=(\\S+)",
               "order": "dstuser"
            }
         },
         {
            "file": "0320-sudo_decoders.xml",
            "path": "ruleset/decoders",
            "name": "sudo2-fields",
            "position": 12,
            "status": "enabled",
            "details": {
               "parent": "sudo",
               "regex": "COMMAND=(\\.+)",
               "order": "command"
            }
         }
      ],
      "totalItems": 2
   }
}
# curl -u foo:bar -k -X GET "http://127.0.0.1:55000/decoders?pretty&q=name~sudo2&limit=1"
{
   "error": 0,
   "data": {
      "items": [
         {
            "file": "0320-sudo_decoders.xml",
            "path": "ruleset/decoders",
            "name": "sudo2-fields",
            "position": 7,
            "status": "enabled",
            "details": {
               "parent": "sudo",
               "program_name": null,
               "regex": "(\\S+)\\s+:\\s+HOST",
               "order": "srcuser",
               "fts": "name,srcuser,location",
               "ftscomment": "First time user executed the sudo command"
            }
         }
      ],
      "totalItems": 6
   }
}
```

```javascript
# mocha test/test_decoders.js 


  Decoders
    GET/decoders
      ✓ Request (312ms)
      ✓ Pagination (277ms)
      ✓ Retrieve all elements with limit=0 (288ms)
      ✓ Sort (287ms)
      ✓ Search (301ms)
      ✓ Filters: File (278ms)
      ✓ Filters: Path (275ms)
      ✓ Filters: query 1 (282ms)
      ✓ Filters: query 2 (281ms)
      ✓ Filters: query 3 (294ms)
      ✓ Filters: Invalid filter
      ✓ Filters: Invalid filter - Extra field
...
27 passing (7s)
```

## TO DO

- Unit tests.
